File: C++/BubbleSort.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

void print_array(auto array)
{
    for(int i : array)
    {
        cout << i << " ";
    }
    cout << endl;
}

void bubble_sort(auto &array)
{
    auto len = array.size();
    int k = 0;
    for (int i = 0; i < len; i++)
    {
        bool flag = false;
        for (int j = len - 1; j > i; j--)
        {
            if (array[j] < array[j-1])
            {
                flag = true;
                auto temp = array[j-1];
                array[j-1] = array[j];
                array[j] = temp;
            }
        }
        
        if(not flag)
        {
            break;
        }

        cout << "after " << ++k << " change: ";
        print_array(array);
    }
}

int main()
{  
    vector<int> array;
    char ch;
    cout << "please input the array(split by space):";
    while((ch = cin.get()) != '\n')
    {
        if(ch != ' ')
            array.push_back(int(ch - '0'));
    }
    cout << "origin is: " ;
    print_array(array);

    bubble_sort(array);
    cout << "result is: " ;
    print_array(array);

    system("pause");
    return 0;
}

File: python/BubbleSort.py
def bubble_sort_ascended(a):
    k =0
    for i in range(0, len(a)):
        flag = False
        k = k + 1
        for j in range(len(a)-1, i, -1):
            if(a[j] < a[j-1]):
                flag = True
                a[j], a[j-1] = a[j-1], a[j]
        if(not flag):
            break
        print("after %d change: "%k, a)
    return a

def bubble_sort_dascended(a):
    k =0
    for i in range(0, len(a)):
        flag = False
        k = k + 1
        for j in range(len(a)-1, i, -1):
            if(a[j] > a[j-1]):
                flag = True
                a[j], a[j-1] = a[j-1], a[j]
        if(not flag):
            break
        print("after %d change: "%k, a)
    return a

def main():
    Array = list(map(int,input("please input the array(split by space):").split()))
    print("input is: ", Array)
    SortedArrayAscend = bubble_sort_ascended(Array)
    print("ascend result is: ", SortedArrayAscend)

    SortedArrayDascend = bubble_sort_dascended(Array)
    print("dascend result is: ", SortedArrayDascend)

if __name__ == "__main__":
    main()